### PR TITLE
update rust to 1.77.2

### PR DESCRIPTION
**Pull request by Xaeroxe**
This also adds a rust-toolchain.toml file, something that our dev and build environments can use to discern the correct version to compile with. If the specified toolchain is missing, rustup will automatically install it.

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -28,24 +28,13 @@ jobs:
...
     - name: Build
-      uses: actions-rs/cargo@v1
-      with:
-        command: build
-        args: --verbose --workspace --exclude xilinx --features srt/async ${{ matrix.args }}
+      run: cargo build --verbose --workspace --exclude xilinx --features srt/async ${{ matrix.args }}
     - name: Lint
       run: cargo clippy --workspace --exclude xilinx --features srt/async --all-targets ${{ matrix.args }} -- --deny warnings
     - name: Test
-      uses: actions-rs/cargo@v1
```
Comment (scottlamb) [inline .github/workflows/rust.yml:45]: This was nice:

> Warnings and errors issued by cargo will be displayed in GitHub UI

Is there a direct replacement somewhere?

File: x264/src/encoder.rs [hunk 2]
```diff
@@ -244,7 +244,7 @@ mod test {
 
         assert_eq!(encoded_frames, 90);
         assert!(encoded.len() > 5000, "encoded.len() = {}", encoded.len());
-        assert!(encoded.len() < 10000, "encoded.len() = {}", encoded.len());
```
Comment (scottlamb) [inline x264/src/encoder.rs]: Was this flaky before? or did it actually change with the Rust upgrade somehow?

File: xcoder/xcoder-quadra/src/decoder.rs [hunk 3]
```diff
@@ -371,51 +371,53 @@ impl<I, E> Drop for XcoderDecoder<I, E> {
 }
 
 #[cfg(test)]
-pub mod test {
```
[the remark this answers is not in the document]
Comment (Xaeroxe) [reply xcoder/xcoder-quadra/src/decoder.rs:374]: `read_frames` is still under a `#[cfg(test)]`, so it won't show up in a non-test build.

Yes this was due to shadowing warnings.

File: x264/src/encoder.rs (hunk 2 above)
Comment (Xaeroxe) [reply x264/src/encoder.rs]: Unclear. The value I was getting in CI was ~16,000. I figured a 60% improvement from a compiler upgrade was possible, so I bumped the ceiling.

File: .github/workflows/rust.yml (hunk 1 above)
Comment (Xaeroxe) [reply .github/workflows/rust.yml:45]: Unfortunately, no. 

So I'd be tempted to keep `actions-rs` in order to retain that. My fear with doing so is that eventually GitHub is going to bring down the hammer and remove node16 entirely. [Based on a previous announcement](https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/), that could be happening very soon. So we might get forced off of `actions-rs` regardless.

File: x264/src/encoder.rs (hunk 2 above)
Comment (scottlamb) [reply x264/src/encoder.rs]: Larger encodings are not an improvement (although they may provide higher video quality), and given that the encoding is done in hardware, shouldn't be caused by the compiler upgrade. If it is, something is really wrong with this test or the production code!

Comment (Xaeroxe) [reply x264/src/encoder.rs]: Okay, right. I had misread the test and thought `len()` was a frame count, not a byte count. The strange part is that this PR doesn't actually change anything of substance for `x264`. I'll try repeating the test without the changes from this PR.

File: .github/workflows/rust.yml (hunk 1 above)
Comment (Xaeroxe) [reply .github/workflows/rust.yml:45]: Actually [we have a date for that now](https://github.blog/changelog/2024-03-07-github-actions-all-actions-will-run-on-node20-instead-of-node16-by-default/)

File: x264/src/encoder.rs (hunk 2 above)
Comment (Xaeroxe) [reply x264/src/encoder.rs]: The commit which bumped this ceiling has been removed.

File: .github/workflows/rust.yml (hunk 1 above)
Comment (scottlamb) [reply .github/workflows/rust.yml:45]: I vote for one idea per PR; maybe trying `FORCE_JAVASCRIPT_ACTIONS_TO_NODE20=true` in a separate PR.

Comment (Xaeroxe) [reply .github/workflows/rust.yml]: Alright, removed. I'll address that in another PR.

File: x264/src/encoder.rs (hunk 2 above)
Comment (Xaeroxe) [reply x264/src/encoder.rs]: @scottlamb The latest test run can no longer reproduce this issue. So either it was caused by one of the commits I removed in response to https://github.com/sportsball-ai/av-rs/pull/193#discussion_r1567938110, or the issue is intermittent.

**Review by scottlamb — APPROVED** [review]

File: x264/src/encoder.rs (hunk 2 above)
Comment (Xaeroxe) [reply x264/src/encoder.rs]: Figured it out https://github.com/sportsball-ai/av-rs/pull/195/commits/f1057e95c2d18ffe17e64220bfcc976fdb97f123

**Pull request merged**